### PR TITLE
Elevate Users to App Wide

### PR DESCRIPTION
**Pull request by orangewolf**
Before reviewing this PR please help get #1640 merged. This PR chains off of that as this code had all the Rubocop fixes in it already, making merging easier.

Move users out of a given tenant and make them site wide. This change leaves the roles to be tenant by tenant, but makes it possible to administer the users at the super admin level, makes single sign on possible and is a stepping stone to a more solid permissions structure.  This code was provided to Hyku by PALNI/PALCI.

**Please note** - this is a breaking change as no migration of tenant specific users to the public schema is currently provided. After a survey of known Hyku installations it was decided that we could move forward with this change with out such a migration simply because no installs came forward that felt they could upgrade to Hyku 3.0 directly and had users in separate tenants. This is especially true because Hyku installs implementing external SSO are likely unaffected by the user move.

@samvera/hyku-code-reviewers

### CODE REVIEW

**Comment by cjcolvar** [comment]
@orangewolf If you don't mind, I'm going to wait until #1640 has been merged and this has been rebased before reviewing to make it easier to work through.

**Comment by orangewolf** [comment]
@cjcolvar totally - that was my main motivator for splitting them up! one should not have to wade through Rubocop changes to see this work =-D

**Comment by cjcolvar** [comment]
@orangewolf Sorry for being slow getting around to this.  It looks like it still needs a rebase since the rubocop commits are hanging around.  I'll start looking through the last three commits (before the merge commit).

File: app/controllers/proprietor/users_controller.rb
Comment (cjcolvar) [inline app/controllers/proprietor/users_controller.rb:81]: Can you explain what the use case for `become` is here?  It seems like persona's become but not exactly.

Comment (orangewolf) [reply app/controllers/proprietor/users_controller.rb:81]: it is the same become as the one in persona - just not as full featured.  There is another PR / feature set that brings persona to the tenant admin section (giving it parity with the Avalon and direct Hyrax version).  We haven't gotten back to consolidating down the proprietor version yet, but that is work I'm hoping to do at some point i the near future.

File: app/models/user.rb
Comment (orangewolf) [reply app/models/user.rb:37]: value comes in from a checkbox and rails checkboxes by default can be nil or "0" for false.  Adding a comment to that affect seems totally reasonable.

File: app/models/user.rb
Comment (orangewolf) [reply app/models/user.rb:71]: the result is the same - that the first user becomes the admin - the difference is that we can't rely on the idea that no other user exists in the tenant, which we could before.

File: app/controllers/hyku/invitations_controller.rb
Comment (orangewolf) [reply app/controllers/hyku/invitations_controller.rb:12]: you're not the only one whose confused - I noticed that when I was pulling the code over to this merge, but convinced my self that it must be right / for a good reason. You're prompting has me looking more deeply at it and I just don't know what I was thinking... and let's be clear - it was me. in 2018 authored that code change with the comment "make invitiations more intutive in new model". I think I must have found the existing code which creates / syncs admin users and felt very clever for re-using it. I do completely agree it seems to only make admins however and that doesn't feel like the right thing to do.

File: app/models/user.rb
Comment (cjcolvar) [reply app/models/user.rb:40]: What about using `ActiveModel::Type::Boolean.new.cast` so this method could support boolean parameter values as well?

File: app/controllers/proprietor/users_controller.rb
Comment (cjcolvar) [reply app/controllers/proprietor/users_controller.rb:81]: How do you feel about dropping `become` from this PR?  Besides not being as built out as in persona, I think it isn't core to the feature being added by this PR and I've gotten the vibe from superiors that they are apprehensive about it without more thought and/or legal review.  (I'll check with them on their reservations when I talk with them about the changes in this PR.) Maybe become could be separated out as another PR so it won't hold up this one for now?

File: app/controllers/hyku/invitations_controller.rb
Comment (orangewolf) [reply app/controllers/hyku/invitations_controller.rb:20]: I don't think so... it seems reasonable that a tenant admin can add other admins. I'm not sure why you'd want to block that.

File: app/views/hyrax/dashboard/_user_activity.html.erb
Comment (orangewolf) [reply app/views/hyrax/dashboard/_user_activity.html.erb:4]: looks like it is resolved in the new version.

File: app/views/hyrax/users/_vitals.html.erb
Comment (orangewolf) [reply app/views/hyrax/users/_vitals.html.erb:8]: added a comment. it was part of a small commit that only fixed the two links. In Hyrax 2.5 they were broken. This was already fixed in master, in a way that didn't seem to back port, so we fixed it with an override so that we didn't diverge.  I've added a comment to this effect.

File: app/views/_admin_util_links.html.erb
Comment (orangewolf) [reply app/views/_admin_util_links.html.erb:11]: good catch, added

File: app/controllers/hyku/invitations_controller.rb
Comment (cjcolvar) [reply app/controllers/hyku/invitations_controller.rb:20]: :+1: I wasn't sure if anyone other than admins could invite users.

File: app/controllers/proprietor/users_controller.rb
Comment (cjcolvar) [inline app/controllers/proprietor/users_controller.rb:76]: Looks like I introduced a rubocop issue with a previous suggestion.  This is hopefully a fix:
```suggestion
      message = if @user.destroy
                  'User was successfully destroyed.'
                else
                  'User could not be destroyed.'
                end
```

File: app/controllers/hyku/invitations_controller.rb
Comment (orangewolf) [reply app/controllers/hyku/invitations_controller.rb:19]: Added

File: app/controllers/account_sign_up_controller.rb
Comment (orangewolf) [reply app/controllers/account_sign_up_controller.rb:19]: yes, if the current user is present, which they may not be. There were a couple bugs here that got worked out by unstacking them until things made sense again.

**Comment by orangewolf** [comment]
@cjcolvar can you mark as resolved any of the above you are ok with? Just want to make sure I don't miss anything. I believe I've gotten through it all again.

**Comment by orangewolf** [comment]
New spec passed locally but not in CI. going to have to look more closely at that (just fyi)

**Comment by cjcolvar** [comment]
@orangewolf The failing CI is the only piece left as far as I can tell.

**Comment by orangewolf** [comment]
```ruby
# this works in Ruby 2.7??? (note the lack of each)
array.split(',') do |a| puts a end
```

Anyway, specs are fixed and were doing their job well.

File: config/locales/en.yml
Comment (cjcolvar) [inline config/locales/en.yml:81]: I found another missing key: `proprietor.users.show.update`
```suggestion
        nav: Users
        show:
          update: Update
```

File: app/services/create_account.rb
Comment (cjcolvar) [inline app/services/create_account.rb:60]: I'm doing some manual testing and this line is throwing a `NoMethodError` because `users` is nil.

File: app/controllers/application_controller.rb
Comment (cjcolvar) [inline app/controllers/application_controller.rb:39]: To resolve the nil I was getting from this method in creating a new tenant account it looks like it needs to return the `users` in case the guard clause evaluates to false:
```suggestion
      users << current_user if current_user && !users.include?(current_user)
      users
```

File: app/services/create_account.rb
Comment (cjcolvar) [reply app/services/create_account.rb:60]: The suggestion in https://github.com/samvera/hyku/pull/1641/files#r482229319 fixed this for me.

**Comment by cjcolvar** [comment]
I think I found a user count inconsistency in the tenant route: http://test.hyku.docker/admin/stats?locale=en#users
I'm seeing 2 instead of the 1 which is reported on http://test.hyku.docker/dashboard?locale=en

File: app/services/create_account.rb
Comment (orangewolf) [reply app/services/create_account.rb:60]: thanks @cjcolvar this makes sense

**Comment by cjcolvar** [comment]
@orangewolf I demo'd this with Ubiquity folks and they think it is a great addition which will help with future work they've been envisioning.  We did come across some odd scenarios that were of concern and since this affects authentication and thus also authorization we want to make sure we get it right from the get go.  The scenarios that we identified all revolve around SSO where a user is able to login to tenant B after registering with tenant A.
1.  With deposit by any user turned on (which it is by default I believe) then a registered user can deposit in any tenant
1.  Any works set to institution level access may be visible by any registered user regardless of tenat
1.  More of an edge case but user profiles are shared between tenants thus not allowing users to have distinct profiles if they have distinct roles within tenant A versus tenant B

I think the first two of these might be able to be solved if we could assign users a "home tenant" maybe in the form of a role stored in the global tenant or inserted into the tenant's role table on registration/login.  Then only users who have the role which matches the tenant they are on can have the deposit and view institution-only visible works (and possibly other abilities).  There could be a feature flipper for this behavior and then there could be discussion over which should be the default behavior.  What do you think about this?  Do you think this should be handled in another PR?  The third scenario can wait until it comes up as a real use case.

There is also interest to have SSO have a feature flipper as well so an installation can opt out of the SSO behavior for non-consorital multi-tenant installations.

I'm happy to provide demos/steps to reproduce if you'd like.

**Comment by orangewolf** [comment]
Hi Chris, we may have discussed this before, but we are doing work right now on PALs which fixes 1 & 2. By straightening out groups, roles and permissions, users will by default not have backend access to tenant dashboards when they don’t have at least one role on that tenant. This fixes the shortcomings of the system in a bunch of ways, including the two you mentioned and several that exist in master. I think this is more flexible than creating home tenants. 

This PR is getting pretty massive, and I think it still solves more than it creates new issues. I’m of a mind to merge it and make tickets to make sure the count issue and 2 permissions issues get solved ASAP. I’m like to be able to show progress at this point before driving too much more time into this step. What do you think? Your collaboration and assistance has been invaluable here.  

The separate profiles things really doesn’t make sense to me - the things in the profile are pretty one to one and it seems like cases where you’d want more than one you’d likely want them tied to separate emails anyway (thus solving the problem)

**Comment by cjcolvar** [comment]
@orangewolf Thanks for the explanation!  I didn't put together what you had said about your team's work on groups/roles with the issues I noted but it totally makes sense now.  It feels good to call it done at this point so :+1:  to merging this now.  

Were you envisioning the groups/roles/permissions work as part of 3.0 or later?  I feel like it is probably a blocker for us but happy to talk about this.

**Review by cjcolvar — APPROVED** [review]
:partying_face: :tada: :confetti_ball:

**Comment by orangewolf** [comment]
I think the whole groups/permissions work is later, but the fix for this issue may be able to come in beforehand. I'll talk to Lea Ann about it today.

**Comment by orangewolf** [comment]
thank you for all your help on this!!

**Pull request merged**
